### PR TITLE
feat: migrate to diagnostic API and support linting on save (feat. golangci-lint)

### LINKED ISSUE #166: Support running linters on `textDocument/didSave`
#69 describes the issues with `golangci-lint`, a linter that doesn't support `stdin` and also needs to run on a full project for context (meaning temp files won't work unless we create them within the project, which could cause other issues). Due to the timing of `textDocument/didChange` notifications, linting files on disk doesn't provide a good experience (the diagnostics "get stuck" and don't update as expected), which is why the PR is on hold. 

One solution would be to allow linters to only run on `textDocument/didSave` notifications. At the moment, it isn't possible to run both types of linters at the same time, since diagnostics from `didChange` linters continually clobber the ones from `didSave` linters, but neovim/neovim#15585 introduces the idea of diagnostic namespaces, which will solve this issue perfectly. 

The question is then whether supporting this is a good idea. The experience between the two types of linters is inconsistent, and I wonder whether it would cause confusion (or cause users to think that null-ls itself is broken). However, it seems like this isn't a problem for the VS Code ecosystem – for example, as far as I can tell the [golangci-lint integration for VS Code](https://golangci-lint.run/usage/integrations/) only runs on save. 

Since there seems to be demand to run these tools, I think it's a good idea to support them once the Neovim PR is merged, as long as we improve the documentation to avoid unnecessary bug reports / complaints. 

Edit: This could also be relevant to #155, though I think using temp files is always a better option if it's possible to do so.

### PR DESCRIPTION
**Pull request by jose-elias-alvarez**
Closes #166 and supersedes #318.

The main change in this PR is to migrate to the `vim.diagnostic` API when available, the biggest advantage of which is the ability to use separate namespaces for each source. This lets us push multiple sets of diagnostics at a time, which enables running linters on save (more on this in a bit) and lays the groundwork for better source management, as mentioned in #337. It's also simpler to use, both internally and from a user config perspective.

Previously, pushing diagnostics on `textDocument/didSave` notifications wasn't possible for the reasons discussed in #166, but with the diagnostic API in place, adding support was simple. The one issue was handling `textDocument/didOpen` notifications, since these should prompt null-ls to run both on-change and on-save linters. I'm not thrilled with the workaround (checking a separate `overrides` table) and want to see if I can come up with something better.

I combined the implementations in #318 and #69 to add support for golangci-lint as the first on-save source. I don't use Go at all and only have a sample repository for testing, so I'd appreciate feedback from @zalegrala and anyone else who's able to test this out in a real development environment. 

Finally, this PR drops support for Neovim version 0.5.0, which is a breaking change but shouldn't be a big deal.

### CODE REVIEW

File: doc/BUILTINS.md [hunk 1]
```diff
@@ -1888,6 +1888,32 @@ local sources = { null_ls.builtins.diagnostics.yamllint }
 - `command = "yamllint"`
 - `args = { "--format", "parsable", "-" }`
 
+### Diagnostics on save
+
+**NOTE**: These sources depend on Neovim version 0.6.0 and are not compatible
```
Comment (zalegrala) [inline doc/BUILTINS.md:1893]: I'm currently running 0.5.1 from Archlinux, but would like to test this out.

Comment (jose-elias-alvarez) [reply doc/BUILTINS.md:1893]: So the issue with 0.5.1 is that diagnostics are correctly generated and on save, but every time you make a change to the buffer, the diagnostics are wiped out until the next save, which (in my opinion) is not a good experience and is inconsistent with how the VS Code extension works. Do you think that's acceptable?

Comment (zalegrala) [reply doc/BUILTINS.md:1893]: That doesn't seem like a great experience, but if what happens when you upgrade is that you get a better experience, then maybe that is acceptable.  I just looked for 0.6.0 source for neovim, but see there are many issues still in the milestone, so not sure on timeline there.

I just checked out the branch and enabled the linter in my config, and what shows up when I run `golangci-lint` manually does not appear in the buffer for me.  Do you think that is related to my version, or perhaps I've not configured something correctly?

Using lunarvim (which I can skip if you think its related) I have this config:
```lua
local linters = require("lvim.lsp.null-ls.linters")
linters.setup({
	 { exe = "golangci_lint", filetypes = { "go" }, },
})
```

Comment (jose-elias-alvarez) [reply doc/BUILTINS.md:1893]: I just pushed a commit disabling the version check so you can try it out. I have no idea about setting up Lunarvim, but all you need to do is pull the latest commit, then run the following code after null-ls is loaded and set up:

```lua
local null_ls = require("null-ls")
null_ls.register(null_ls.builtins.diagnostics.golangci_lint)
```

Comment (zalegrala) [reply doc/BUILTINS.md:1893]: That works.  Now I see what your saying about the buffer change.  If its the best we can do while we wait for 0.6.0 I'd take it, but I do agree it isn't ideal.

Comment (marcomayer) [reply doc/BUILTINS.md:1893]: Ah just found this and that you're already discussing the issue I mentioned in #318. If it's fixed with nvim 0.6 I also think it's a better-than-nothing start 👍

**Comment by David-Else** [comment]
> Finally, this PR drops support for Neovim version 0.5.0, which is a breaking change but shouldn't be a big deal.

Could you make a 0.5.1 compat branch when this happens? They do that with treesitter https://github.com/nvim-treesitter/nvim-treesitter/tree/0.5-compat , you could delete it as soon as 0.6 is out officially :)

**Comment by rafikdraoui** [comment]
Hello! Here is what I experienced after installing this branch.


## 1. Runtime error from golangci-lint isn't handled

If golangci-lint is run in a directory that doesn't have a `go.mod` file, it returns with exit code 5 and the message `ERRO Running error: context loading failed: no go files to analyze`

Opening the Go file in neovim, the following error message is displayed:

```
null-ls: failed to run generator: ...e/pack/packer/start/null-ls.nvim/lua/null-ls/helpers.lua:192: error in generator output: level=error msg="Running error: context loading failed: no go files to analyze"
```



## 2. `params.output.Issues` can be null in `on_output` callback

After adding a `go.mod` file to make golangci-lint happy:

```
module hello

go 1.17
```

The linter now runs, but there is still an error message:

```
null-ls: failed to run generator: ....nvim/lua/null-ls/builtins/diagnostics/golangci_lint.lua:32: bad argument #1 to 'ipairs' (table expected, got userdata)
```

It turns out that `params.output.Issues` can be `vim.NIL` in some cases:

```
[DEBUG Sat 13 Nov 2021 09:18:03 PM] ...ite/pack/packer/start/null-ls.nvim/lua/null-ls/utils.lua:74: output: {"Issues":null,"Report":{"Linters":[...]}}
```

```
$ golangci-lint run --fast --out-format=json | jq .Issues
null
```

This feels like an issue in golangci-lint: ideally it would set `Issues` to an empty list rather than `null` for consistency.

In the mean time, we could have a null-check on `params.output.Issues` before passing it to `ipairs` in https://github.com/jose-elias-alvarez/null-ls.nvim/blob/fe252e54d3c9c7e2a90088e43329035fa6898e55/lua/null-ls/builtins/diagnostics/golangci_lint.lua#L28

**Comment by jose-elias-alvarez** [comment]
> > Finally, this PR drops support for Neovim version 0.5.0, which is a breaking change but shouldn't be a big deal.
> 
> Could you make a 0.5.1 compat branch when this happens? They do that with treesitter https://github.com/nvim-treesitter/nvim-treesitter/tree/0.5-compat , you could delete it as soon as 0.6 is out officially :) If that is annoying some kind of instructions to freeze the version in `vim-plug`/`packer`?

0.5.1 will still be supported until 0.6.0 release (and probably for a couple of months afterwards). This PR just removes 0.5.0 support, since I don't think many users are still on it and accounting for 3 potential diagnostic APIs was too much.

**Comment by jose-elias-alvarez** [comment]
@rafikdraoui Thanks for trying this out and for the detailed feedback! I've addressed the second issue for now by adding a typecheck. 

For the first issue, I believe we have at least a few linters that will error out if the user attempts to run them without a config file. We could either ask users to set a condition for the source that checks for the presence of a config file or do something like what we do with [ESLint](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/on-save-handling/lua/null-ls/builtins/diagnostics/eslint_d.lua), where we inject the error as a diagnostic (a few VS Code extensions have this behavior too). Its presence in this PR is more of a proof-of-concept than anything else, and I'd rather leave it for actual users to improve it in subsequent PRs.

**Comment by jose-elias-alvarez** [comment]
I've gone back and forth on whether to block on-save linters on 0.5.1, and I think @marcomayer's experience shows that users who are unaware of the limitation are likely to conclude that null-ls itself is broken when encountering the behavior. 

Either way, it seems that 0.6 release is [not that far out](https://neovim.discourse.group/t/neovim-0-6-stabilization-period-begins-today-10-31-2021/1329). What I'll do for now is merge this to hopefully get some more feedback on this from 0.6.0 users (who are surely more used to things breaking) and make sure the migration to the diagnostics API is all set for release.

**Pull request merged**